finetune/base_models/bert/featurizer.py

```diff
@@ -140,4 +145,279 @@ def bert_featurizer(
         return output_state
 
 
+def long_doc_bert_featurizer(
+    X,
+    encoder,
+    config,
+    train=False,
+    reuse=None,
+    context=None,
+    total_num_steps=None,
+    underlying_model=BertModel,
+    **kwargs
+):
+    """
+    The transformer element of the finetuning model. Maps from tokens ids to a dense, embedding of the sequence.
+
+    # TODO Do I need to do anything to get static BERT embeddings and then only train the pooling operation?
+
+    :param X: A tensor of token indexes with shape [batch_size, sequence_length, token_idx]
+    :param encoder: A TextEncoder object.
+    :param config: A config object, containing all parameters for the featurizer.
+    :param train: If this flag is true, dropout and losses are added to the graph.
+    :param reuse: Should reuse be set within this scope.
+    :return: A dict containing;
+        features: The output of the featurizer_final state.
+    """
+    bert_config = BertConfig(
+        vocab_size=encoder.vocab_size,
+        hidden_size=config.n_embed,
+        num_hidden_layers=config.n_layer,
+        num_attention_heads=config.n_heads,
+        intermediate_size=config.bert_intermediate_size,
+        hidden_act=config.act_fn,
+        hidden_dropout_prob=config.resid_p_drop,
+        attention_probs_dropout_prob=config.attn_p_drop,
+        # max_position_embeddings=config.max_length,
+        max_position_embeddings=config.max_chunk_length,
+        type_vocab_size=2,
+        initializer_range=config.weight_stddev,
+        low_memory_mode=config.low_memory_mode,
+        pos_injection=config.context_injection,
+        reading_order_removed=config.reading_order_removed,
+        anneal_reading_order=config.anneal_reading_order,
+        positional_channels=config.context_channels,
+    )
+
+    initial_shape = tf.shape(input=X)
+    initial_shape = tf.compat.v1.Print(
+        initial_shape, ["initial_shape", initial_shape], summarize=-1
+    )
+
+    batch_size = initial_shape[0]
+    sequence_len = initial_shape[1]
+
+    # [batch_size, sequence_length]
+    X = tf.reshape(X, shape=tf.concat(([-1], initial_shape[-1:]), 0))
+    X.set_shape([None, None])
+    X = tf.compat.v1.Print(X, ["X", X, tf.shape(X)], summarize=-1)
+
+    # Add padding to X to later break out a separate chunk dimension
+    """
+    FIXME Remove comments with math. Those are with these assumptions for validating
+    padding / chunking logic
+    * config.max_chunk_length = 64
+    * config.batch_size_scaler = 4
+    * sequence_len = 300
```

Comment:
If I understand correctly, this section right now pads out every example in batch to the longest examples number of chunks.  There's likely a way to do this more efficiently by merging the batch + chunk dims but undoing the operation and re-padding before the aggregation op, but definitely falls into the "future work" category.

Reply:
It's a little more than that because of the `batch_size_scaler`. Basically for the dims to work out, the number of chunks needs to be divisible by the batch_size_scaler. So if the max sequence length in a batch is 300 with a chunk length of 64 and batch_size_scaler of 4, you need to pad out to 512 (512/64=8 chunks is divisible by 4) rather than just 320 (5 chunks)

Reply:
I guess my point was more -- in this section of code all chunks are created equal.  We just want to apply some operation to many chunks from (possibly) many examples.  Later on the example that each chunk belongs to becomes important.  But in this portion of the code, it could be possible to merge the batch and chunk dims into a single sequence, map the embedding fn across that sequence, and then later reorganize chunks by example.

Reply:
Problem with this is we know we likely can't fit batch size * num chunks though the featurizer simultaniously. And doing it one by one is not efficient. I think if we wanted to further optimise this - something similar to [this](https://github.com/tensorflow/tensor2tensor/blob/5623deb79cfcd28f8f8c5463b58b5bd76a81fd0d/tensor2tensor/utils/expert_utils.py#L582) to allow us to skip over the padding chunks. But I think this is good for now personally